config.go

```diff
@@ -164,12 +166,20 @@ func NewDevelopmentConfig() Config {
 
 // Build constructs a logger from the Config and Options.
 func (cfg Config) Build(opts ...Option) (*Logger, error) {
+	return cfg.BuildWithSinks(DefaultSinks(), opts...)
+}
+
+// BuildWithSinks uses the map of sinks to construct a logger from the
+// passed options. For example, the option "stdout" will construct a logger
+// that routes output to the stdout sink, but "/filepath" will fallback
+// to writing to file.
+func (cfg Config) BuildWithSinks(sm map[string]Sink, opts ...Option) (*Logger, error) {
```

Comment:
This is a nice idea - I'd like to make this portion of zap more extensible. I also like that this approach avoids adding global state.

However, this works very differently from extending encoders: for that, we use a global registry and a top-level `RegisterEncoder` function. For consistency, I'd actually prefer to add another global registry for sinks. (Generally, I much prefer dependency injection, but I think consistency is more important here.)

@prashant and @abhinav, what do you think?

Reply:
I think I'd prefer consistency as well here, and have a global registry for both encoders and sinks.

If we want users to be able to specify the exact registries, we'll likely want to add a single method that accepts both registries rather than a method for each.

Reply:
Thanks for the feedback @akshayjshah @prashantv. I never saw `RegisterEncoder` but after seeing it, your requests make complete sense. I have since added a commit with `RegisterSink` to mimc that functionality. If @abhinav feels strongly about reverting back to the dependency injection style, no worries, we can just revert one commit.